### ISSUE TITLE
Bytes released to OS still show up in use

### ISSUE THREAD

**Issue by santhoshmprabhu** [gperftools/gperftools#1092]
I am running into an issue where memory reported as "released to OS" still appears to be in use.
I am running google-perftools 2.5 on Ubuntu 18.04, with Transparent Huge Pages disabled. Here is my testing code (written in google test framework)
```
TEST(Deserialization, TestLoad) {
    {
        Datatype object;
        const string sourceRoot = string(::getenv("SOURCE_ROOT"));
        Deserializer::deserialize(&object, sourceRoot + "/infile");
        cout<<"Stats before deletion: "<<endl;
        malloc_stats();
    }
    cout<<"Stats after deletion: "<<endl;
    malloc_stats();
    MallocExtension::instance()->ReleaseFreeMemory();
    cout<<"Stats after ReleaseFreeMemory: "<<endl;
    malloc_stats();
    //Check memory here
    getchar();
}
```
This produces the following output:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Deserialization
[ RUN      ] Deserialization.TestLoad
Stats before deletion: 
------------------------------------------------
MALLOC:    20655515608 (19698.6 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +     68000752 (   64.9 MiB) Bytes in central cache freelist
MALLOC: +            0 (    0.0 MiB) Bytes in transfer cache freelist
MALLOC: +      2120760 (    2.0 MiB) Bytes in thread cache freelists
MALLOC: +    110784704 (  105.7 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =  20836421824 (19871.2 MiB) Actual memory used (physical + swap)
MALLOC: +       327680 (    0.3 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  20836749504 (19871.5 MiB) Virtual address space used
MALLOC:
MALLOC:        1763358              Spans in use
MALLOC:              1              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
Stats after deletion: 
------------------------------------------------
MALLOC:         185856 (    0.2 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +     69866344 (   66.6 MiB) Bytes in central cache freelist
MALLOC: +     22938640 (   21.9 MiB) Bytes in transfer cache freelist
MALLOC: +      5902984 (    5.6 MiB) Bytes in thread cache freelists
MALLOC: +    110784704 (  105.7 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    209678528 (  200.0 MiB) Actual memory used (physical + swap)
MALLOC: +  20627070976 (19671.5 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  20836749504 (19871.5 MiB) Virtual address space used
MALLOC:
MALLOC:          12754              Spans in use
MALLOC:              1              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
Stats after ReleaseFreeMemory: 
------------------------------------------------
MALLOC:         185856 (    0.2 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +     69866344 (   66.6 MiB) Bytes in central cache freelist
MALLOC: +     22938640 (   21.9 MiB) Bytes in transfer cache freelist
MALLOC: +      5902984 (    5.6 MiB) Bytes in thread cache freelists
MALLOC: +    110784704 (  105.7 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    209678528 (  200.0 MiB) Actual memory used (physical + swap)
MALLOC: +  20627070976 (19671.5 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  20836749504 (19871.5 MiB) Virtual address space used
MALLOC:
MALLOC:          12754              Spans in use
MALLOC:              1              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
```
`top` output at the point where the program waits for input:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                 
 6499 santhosh  20   0 19.484g 0.019t   9492 S   0.0 30.9   3:15.49 Deserialization
```

I'd appreciate any input.

**Comment by sanpra1989** [comment]
Part of `/cat/pid/smaps`
```564d47b7a000-56522214c000 rw-p 00000000 00:00 0                          [heap]
Size:           20354888 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:            20338592 kB
Pss:            20338592 kB
Shared_Clean:          0 kB
Shared_Dirty:          0 kB
Private_Clean:  20135000 kB
Private_Dirty:    203592 kB
Referenced:       203592 kB
Anonymous:      20338592 kB
LazyFree:       20135000 kB
AnonHugePages:         0 kB
ShmemPmdMapped:        0 kB
Shared_Hugetlb:        0 kB
Private_Hugetlb:       0 kB
Swap:                  0 kB
SwapPss:               0 kB
Locked:                0 kB
VmFlags: rd wr mr mw me ac sd 
```

**Comment by alk** [comment]
Thanks for reporting the issue. One thing that is notable about your use-case is that MADV_FREE is being used which we do not enable by default. I.e. looking at LazyFree figures above.

Smaps figures posted above, clearly show that malloc is returning this memory. Via MADV_FREE, but since it it "lazy" it is not actually immediately removed from process RSS.

So how exactly did you get this build of tcmalloc? Asking because I don't feel comfortable about MADV_FREE exactly because of cases like yours.

Still even with that something is not quite right with such massive virtual address space growth. If can extract your use case into small test program that reproduces this growth, please post it.

**Comment by sanpra1989** [comment]
This library was installed by the default `libgoogle-perftools-dev` package on Ubuntu 18.04. I am not familiar with the use of MADV_FREE and how it impacts the collection of memory, but do you think I should build tcmalloc from source? I could do so if that would help.

```dpkg -S /usr/lib/x86_64-linux-gnu/libtcmalloc.so
libgoogle-perftools-dev: /usr/lib/x86_64-linux-gnu/libtcmalloc.so```

The original program that I used cannot be released. I am now trying to reproduce the issue on something I could post publicly.

**Comment by alk** [comment]
So apparently some time ubuntu shipped gperftools with MADV_FREE thingy. I just tested 22.04 and it is fine.

**Comment by alk** [comment]
Ah, it was our problem. We "automagically" used MADV_FREE when available. But not anymore. Fixed in gperftools 2.6.
